refactor: Add support for extra transaction signature validation

ximinez opened the pull request:
## High Level Overview of Change

- Restructures `STTx` signature checking code to be able to handle a `sigObject`, which may be the full transaction, or may be an object field containing a separate signature. Either way, the `sigObject` can be a single- or multi-sign signature.
- This is distinct from 550f90a75e (#5594), which changed the check in `Transactor`, which validates whether a given account is allowed to sign for the given transaction. This cryptographically checks the signature validity.

Because nothing actually uses the new `sigObject` capability, this can be considered a refactor.

### Context of Change

This change was originally part of #5270 (Lending Protocol), but @tequdev requested to make it available sooner so it could be used in another project.

### Type of Change

- [X] Refactor (non-breaking change that only restructures code)

ximinez committed:
Add support for extra transaction signature validation

- Restructures `STTx` signature checking code to be able to handle
  a `sigObject`, which may be the full transaction, or may be an object
  field containing a separate signature. Either way, the `sigObject` can
  be a single- or multi-sign signature.
- This is distinct from 550f90a75e (#5594), which changed the check in
  Transactor, which validates whether a given account is allowed to sign
  for the given transaction. This cryptographically checks the signature
  validity.

ximinez committed:
Add jtx, STObject, and RPC support for sig object fields

ximinez committed:
Merge branch 'develop' into ximinez/lending-sttx-checksign

ximinez committed:
Merge branch 'develop' into ximinez/lending-sttx-checksign

ximinez committed:
Merge branch 'develop' into ximinez/lending-sttx-checksign

ximinez committed:
Merge branch 'develop' into ximinez/lending-sttx-checksign

ximinez committed:
Merge branch 'develop' into ximinez/lending-sttx-checksign

a1q123456 commented on include/xrpl/protocol/STObject.h:
```diff
@@ -244,6 +244,9 @@ class STObject : public STBase, public CountedObject<STObject>
     getFieldV256(SField const& field) const;
+    // If not found, returns an object constructed with the given field
+    STObject
+    getFieldObject(SField const& field) const;
```
I think the name getFieldObject is technically clear, but it hides an important nuance: it constructs a new object if the field doesn’t exist, which is unexpected behaviour for a “get” method. That violates the principle of least surprise.

I'd name it `getOrCreateFieldObject`

ximinez replied:
> I think the name getFieldObject is technically clear, but it hides an important nuance: it constructs a new object if the field doesn’t exist, which is unexpected behaviour for a “get” method. That violates the principle of least surprise.
> 
> I'd name it `getOrCreateFieldObject`

`getFieldObject` follows the same pattern as all of the other `getField...` functions. That's just how `STObject` works.

a1q123456 commented on include/xrpl/protocol/STTx.h:
```diff
@@ -88,7 +88,13 @@ class STTx final : public STObject, public CountedObject<STTx>
     // Outer transaction functions / signature functions.
     Blob
-    getSignature() const;
+    getSignature(STObject const& sigObject) const;
```
I think this should be static if it's not related to `this`

ximinez replied:
> I think this should be static if it's not related to `this`

Yeah, that makes sense. Will do.

a1q123456 commented on src/libxrpl/protocol/STTx.cpp:
```diff
@@ -234,37 +234,71 @@ STTx::getSeqValue() const
+STTx::sign(
+    PublicKey const& publicKey,
+    SecretKey const& secretKey,
+    std::optional<std::reference_wrapper<SField const>> signatureTarget)
```
`std::optional<std::reference_wrapper<SField const>>`  vs `SField const*`
I'm not sure which is better, but I think we should be consistent in this function and `STTx::checkSign`. So we should either use a raw pointer or `std::optional<std::reference_wrapper<T>>` in both `STTx::sign` and `STTx::checkSign`.


But yeah, I guess we can get rid of `reference_wrapper` after we upgrade to c++ 23.

ximinez replied:
> `std::optional<std::reference_wrapper<SField const>>` vs `SField const*` I'm not sure which is better, but I think we should be consistent in this function and `STTx::checkSign`. So we should either use a raw pointer or `std::optional<std::reference_wrapper<T>>` in both `STTx::sign` and `STTx::checkSign`.

I implemented the two functions differently based on what's easier for me to write. `SField` lookups for `sign` return `SField const&`. Whereas the `STObject` passed to `checkSign` is `this` more often than not, which is a ptr, obviously.

What I realize from your question, though is that the overload of `checkSign` that takes the ptr, does not need to be exposed. So what I'm going to do is make it `private`, change it to take a `STObject const&`, and change the (internal) callers to pass `*this` instead of `nullptr`.

a1q123456 replied:
Yeah it makes sense to pass by const& instead of a pointer if it's not optional. Thanks, Ed!

a1q123456 commented on src/libxrpl/protocol/STTx.cpp:
```diff
@@ -234,37 +234,71 @@ STTx::getSeqValue() const
 }
 
+Expected<void, std::string>
+STTx::checkSign(
```
I think it'll be clearer if we call `checkSign` twice instead of trying to merge the two paths together. So instead of having another overload and a bit twisted logic, we could just call `checkSign` twice:
```c++
checkSign(requireCanonicalSig, rules, nullptr);
if (isFieldPresent(sfCounterpartySignature))
    {
        auto const counterSig = getFieldObject(sfCounterpartySignature);
        if (auto const ret = 
checkSign(requireCanonicalSig, rules, &counterSig);
            !ret)
            return Unexpected("Counterparty: " + ret.error());
    }
```

Also it seems that `checkSign` has been modified so that it checks against `pSig` if it's not nullptr and it checks against `this` otherwise.

I think it'd be better to just construct a new STTx and call checkSign directly.

```c++
checkSign(requireCanonicalSig, rules, nullptr);
if (isFieldPresent(sfCounterpartySignature))
    {
        auto const counterSig = STTx{getFieldObject(sfCounterpartySignature)};
        if (auto const ret = 
checkSign(requireCanonicalSig, rules, &counterSig);
            !ret)
            return Unexpected("Counterparty: " + ret.error());
        if (auto const ret = counterSig.checkSign(requireCanonicalSig, rules);
            !ret)
            return Unexpected("Counterparty: " + ret.error());
    }
```
and revert changes related to `pSig` in `checkSignleSign` and `checkMultiSign` and rename this method to something that describes the use case better.

ximinez replied:
> I think it'll be clearer if we call `checkSign` twice instead of trying to merge the two paths together. So instead of having another overload and a bit twisted logic, we could just call `checkSign` twice:

I'm confused. I am calling `checkSign` twice. (Well, it'll be twice when the counterpartysignature block is uncommented.)


> Also it seems that `checkSign` has been modified so that it checks against `pSig` if it's not nullptr and it checks against `this` otherwise.
> 
> I think it'd be better to just construct a new STTx and call checkSign directly.

This _can't_ work. A signature object like `sfCounterpartySignature` is not a whole `STTx`. It's just the signature fields, which have signed the contents of the actual `STTx`. 

Sloppy metaphor: Imagine you write a report that contains a bunch of data, analysis, computations, etc. You print that report and write your physical signature on it. As a peer review, I get that printed report, do whatever verification I need to ensure that it's accurate, and then I seal it in an envelope, and sign the envelope. The letter is the `STTx`. The envelope is the `CounterpartySignature`. The envelope is meaningless without the letter inside of it.

a1q123456 replied:
That's a really nice explanation! Thanks, Ed!

a1q123456 commented on src/libxrpl/protocol/STTx.cpp:
```diff
@@ -588,11 +637,12 @@ STTx::getBatchTransactionIDs() const
-
-    for (STObject const& rb : getFieldArray(sfRawTransactions))
-        batch_txn_ids_.push_back(rb.getHash(HashPrefix::transactionID));
+    // Don't early return so that the size assert is always hit.
```
Or we can move the assert before the if?

ximinez replied:
> Or we can move the assert before the if?

No, because the vector is being populated inside of the `if`. The condition may be false before it.

The reason the comment is written that way is because the original code was returning the list if it wasn't empty, thus bypassing the assert. I wanted the assert to be checked on every call to this function. The other option would have been to duplicate the assert, which would be dumb.

I rewrote the comment to be more clear about what the code is doing without depending on knowing what was there before.

a1q123456 replied:
Was thinking about checking against `getFieldArray(sfRawTransactions)` but yeah I think it's the same. Thanks, Ed!

ximinez committed:
Review feedback from @a1q123456

- Convert STTx::getSignature to be static.
- Make the STTx::checkSign that takes a signature object private, and
  change it and all the other relevant functions to take the signature
  object as a `const&`.
- Change `getBatchTransactionIDs` to return a `const&` to reduce the
  number of vector copies. Rename `batchTxnIds_` to CamelCase. Update
  the text of the internal comment.

ximinez committed:
Merge branch 'develop' into ximinez/lending-sttx-checksign

a1q123456 reviewed (approved):
LGTM

ximinez committed:
Merge branch 'develop' into ximinez/lending-sttx-checksign

ximinez committed:
Merge branch 'develop' into ximinez/lending-sttx-checksign

ximinez committed:
Merge branch 'develop' into ximinez/lending-sttx-checksign

shawnxie999 commented on src/xrpld/rpc/detail/TransactionSign.cpp:
are we committed to support this RPC? the "sign" RPC has been deprecated IIRC

ximinez replied:
> are we committed to support this RPC? the "sign" RPC has been deprecated IIRC

It's still used quite heavily in test environments. That was the main reason I had to update `sign`.

shawnxie999 reviewed (approved):
Looks good 👍

ximinez committed:
Merge branch 'develop' into ximinez/lending-sttx-checksign

ximinez committed:
Merge branch 'develop' into ximinez/lending-sttx-checksign

bthomee committed:
Merge branch 'develop' into ximinez/lending-sttx-checksign

ximinez committed:
Merge branch 'develop' into ximinez/lending-sttx-checksign

ximinez committed:
Merge branch 'develop' into ximinez/lending-sttx-checksign

ximinez committed:
Merge branch 'develop' into ximinez/lending-sttx-checksign

bthomee committed:
Merge branch 'develop' into ximinez/lending-sttx-checksign
